ic-agent/src/agent/http_transport/dynamic_routing/dynamic_route_provider.rs

```diff
@@ -0,0 +1,596 @@
+//! An implementation of the [`RouteProvider`](crate::agent::http_transport::route_provider::RouteProvider) for dynamic generation of routing urls.
+
+use std::{
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+use anyhow::anyhow;
+use arc_swap::ArcSwap;
+use candid::Principal;
+use reqwest::Client;
+use tokio::{
+    sync::{mpsc, watch},
+    time::timeout,
+};
+use tokio_util::{sync::CancellationToken, task::TaskTracker};
+use tracing::{error, info, warn};
+use url::Url;
+
+use crate::{
+    agent::http_transport::{
+        dynamic_routing::{
+            health_check::{HealthCheck, HealthChecker, HealthManagerActor},
+            messages::FetchedNodes,
+            node::Node,
+            nodes_fetch::{Fetch, NodesFetchActor, NodesFetcher},
+            snapshot::routing_snapshot::RoutingSnapshot,
+            type_aliases::GlobalShared,
+        },
+        route_provider::RouteProvider,
+    },
+    AgentError,
+};
+
+pub const IC0_SEED_DOMAIN: &str = "ic0.app";
+
+const MAINNET_ROOT_SUBNET_ID: &str =
+    "tdb26-jop6k-aogll-7ltgs-eruif-6kk7m-qpktf-gdiqx-mxtrf-vb5e6-eqe";
+
+const FETCH_PERIOD: Duration = Duration::from_secs(5);
+const FETCH_RETRY_INTERVAL: Duration = Duration::from_millis(250);
+const TIMEOUT_AWAIT_HEALTHY_SEED: Duration = Duration::from_millis(1000);
+const HEALTH_CHECK_TIMEOUT: Duration = Duration::from_secs(2);
+const HEALTH_CHECK_PERIOD: Duration = Duration::from_secs(1);
+
+const DYNAMIC_ROUTE_PROVIDER: &str = "DynamicRouteProvider";
+
+#[derive(Debug)]
+pub struct DynamicRouteProvider<S> {
+    fetcher: Arc<dyn Fetch>,
+    fetch_period: Duration,
+    fetch_retry_interval: Duration,
+    checker: Arc<dyn HealthCheck>,
+    check_period: Duration,
+    snapshot: GlobalShared<S>,
+    tracker: TaskTracker,
+    seeds: Vec<Node>,
+    token: CancellationToken,
+}
+
+impl<S> RouteProvider for DynamicRouteProvider<S>
+where
+    S: RoutingSnapshot + 'static,
+{
+    fn route(&self) -> Result<Url, AgentError> {
+        let snapshot = self.snapshot.load();
+        let node = snapshot.next().ok_or_else(|| {
+            AgentError::RouteProviderError("No healthy API nodes found.".to_string())
+        })?;
+        Ok(node.to_routing_url())
+    }
+}
+
+impl<S> DynamicRouteProvider<S>
+where
+    S: RoutingSnapshot + 'static,
+{
+    pub fn new(snapshot: S, seeds: Vec<Node>, http_client: Client) -> Self {
```

Comment:
This means that to create an agent with the discovery library, you need to do something like:
```
let snapshot = LatencyBasedSnapshot::new();
let http_client = Client::builder().build().unwrap();
let route_provider = HealthCheckRouteProvider::new(
            snapshot,
            vec![Node::new(domain_seed)],
            http_client
);
let transport = ReqwestTransport::create_with_client_route(
        Arc::clone(&route_provider) as Arc<dyn RouteProvider>,
        http_client,
    )
    .expect("failed to create transport");
    
let agent = Agent::builder()
        .with_transport(transport)
        .build()?;
```

Can't we do something like and just provide defaults:
```
let agent = Agent::builder()
        .with_latency_based_routing()
        .build()?;
```
or
```
let agent = Agent::builder()
        .with_health_based_routing()
        .build()?;
```

Reply:
If someone wants not the default, they can then go through the process above and create their own snapshot/route provider, etc.

Reply:
Eventually this is the direction to go, yes! But i don't want to change the `agent` during the library migration phase., better in the follow up.